app/src/org/commcare/activities/connect/ConnectIdSsoHelper.java

```diff
@@ -70,7 +70,24 @@ public static AuthInfo.TokenAuth acquireSsoTokenSync(Context context) {
         return hqTokenAuth;
     }
 
-    private static AuthInfo.TokenAuth retrieveConnectToken(Context context) {
+    public static void retrieveConnectTokenAsync(Context context, TokenCallback callback) {
```

Comment:
looks like what's happening here is - `Fragment(UI) -> getConnectOpportunities(UI) -> retrieveConnectTokenAsync(UI) -> creates a new thread(t) -> callback(t) -> getInternal()(t) -> triggers AsyncTask(t)` . My question is does not an async task needs to get triggered from UI thread ?

Reply:
I changed the code to use an AsyncTask instead, [ae8132a](https://github.com/dimagi/commcare-android/pull/2695/commits/ae8132a361006017d9fa849bb7223ca3482db466)